### ISSUE TITLE
Under prediction of CACS

### ISSUE THREAD

**Issue by koozalli** [Raffi-Hagopian/AI-CAC#6]
Hi Raffi Hagopian,

I have tested your model on non-gated chest CT DICOM images from COCA- Coronary Calcium and chest CT’s Dataset (https://aimi.stanford.edu/datasets/coca-coronary-calcium-chest-ct)

I am also getting a tendency of lower CAC score by AI-CAC than that reported by human experts.

results for different thresholds as analogy to Table 1 in your paper are as follows:
threshold= 1
tn=73, fp=10, fn=31, tp=80
accuracy=(tp+tn)/(tn+fp+fn+tp)=0.788659793814433
sensitivity=tp/(tp+fn)=0.7207207207207207
specificity=tn/(tn+fp)=0.8795180722891566
F1=(2*tp)/(2*tp+fp+fn)=0.7960199004975125

---------------------------------------------------------------------
threshold= 100
tn=134, fp=4, fn=23, tp=33
accuracy=(tp+tn)/(tn+fp+fn+tp)=0.8608247422680413
sensitivity=tp/(tp+fn)=0.5892857142857143
specificity=tn/(tn+fp)=0.9710144927536232
F1=(2*tp)/(2*tp+fp+fn)=0.7096774193548387

----------------------------------------------------------------------
threshold= 400
tn=164, fp=4, fn=10, tp=16
accuracy=(tp+tn)/(tn+fp+fn+tp)=0.9278350515463918
sensitivity=tp/(tp+fn)=0.6153846153846154
specificity=tn/(tn+fp)=0.9761904761904762
F1=(2*tp)/(2*tp+fp+fn)=0.6956521739130435

-----------------------------------------------------------------------
and a confusion matrix as analogy to Figure 3.D in your paper is here: [AI-CAC-results-on-COCA-stanford-dataset.pdf](https://github.com/user-attachments/files/26777983/AI-CAC-results-on-COCA-stanford-dataset.pdf), where _True label_ is for the CAC reported by the dataset, and _Predicted label_ is for the CAC reported by AI-CAC.

Other than the following two modifications, all the source code is same:
-I am using the latest version of MONAI, in which _SwinUNETR_ doesnt accept an argument _img_size_, so I am simply removing this argument [here](https://github.com/Raffi-Hagopian/AI-CAC/blob/2d912007ea97924ceb4ee6fc0766174261eba38e/main_inference.py#L66).
-according to [this line in filter_series.py](https://github.com/Raffi-Hagopian/AI-CAC/blob/2d912007ea97924ceb4ee6fc0766174261eba38e/filter_series.py#L77) a series will be removed if it doesnt have _study_desc_ or _series_desc_. Since _SeriesDescription_ and _StudyDescription_ fields are _None_ in all the DICOM files in the dataset, I have commented out [this line](https://github.com/Raffi-Hagopian/AI-CAC/blob/2d912007ea97924ceb4ee6fc0766174261eba38e/filter_series.py#L100) and the line next to it.

The field of view is same as the sample images in your paper, see those images from the dataset as an example: <img width="800" height="800" alt="Image" src="https://github.com/user-attachments/assets/d3c5609f-5650-4226-8bd8-121d13d612cf" /> <img width="800" height="800" alt="Image" src="https://github.com/user-attachments/assets/899dd485-6a91-42c9-b68b-28c0bc6c4ca5" />

you can also see the code that I have used to run the model, and the code that I have used to generate the confusion matrix here:
- https://www.kaggle.com/code/rabeasaleh2/ai-cac-test
- https://www.kaggle.com/code/rabeasaleh2/matplotlib (last 5 cells)

What might be the reason of the lower performance on this dataset in your opinion?

Thanking you

**Comment by Raffi-Hagopian** [comment]
Thank you for doing this detailed analysis! Here are my thoughts on why the model might be under-calling on this dataset and what you can check:

1) **Slice thickness distribution in the data:** I think the COCA non-gated dataset skews heavily towards the thicker end (all ~5mm?), Gated CAC Agatston scoring usually done on 3mm slices. If the non-gated scan thickness distribution is skewed to thicker slices than 3mm it will result in more under-calling CAC due to partial volume averaging . If it is thinner it will overcall CAC. 
2) **2D slice dimensions:** You mentioned that the img_size parameter was removed, hopefully it's still being processed at 512x512 dimensions (resampling to lower dimensions can cause lower performance). 
3) **Scanner Manufacturer Differences:** Our data was mostly from GE/Phillips/Siemens/Canon[Toshiba]
4) **Population Differences:** VA veteran population skews ~90% male vs. the Stanford university population

Let me know what you find! I don't remember if this was their testing dataset, but if so the F1s still look higher with our model from what I think was reported in the original paper's non-gated vs. gated comparisons.

**Comment by koozalli** [comment]
Thank you very much for your reply.

1.[dicom_input_one_series.csv](https://github.com/user-attachments/files/26848993/dicom_input_one_series.1.csv)
Out of the 213 patients in the non-gated dataset of COCA, 194 patients are remaining after filtering by _filter_dicom_df_.
all the 194 patients have slice thickness of 5 mm, except five patients. all of those five patients are correctly classified if we consider the four categories (0, 1-100, 101-400, >400). the details and scores for those five patients are here:[patients-with-slice-thickness-less-than-5.xlsx](https://github.com/user-attachments/files/26849021/patients-with-slice-thickness-less-than-5.xlsx) .

2.I have checked the slice dimensions by printing _batch.shape_ just before [this line](https://github.com/Raffi-Hagopian/AI-CAC/blob/2d912007ea97924ceb4ee6fc0766174261eba38e/main_inference.py#L97) _batch_out = model(batch.float()) #might need to move batch dimension_. all spatial dimensions are (512,512).
I have also printed _h, w, z_length_ after [this line](https://github.com/Raffi-Hagopian/AI-CAC/blob/2d912007ea97924ceb4ee6fc0766174261eba38e/dataset_generator_inference.py#L36) _h, w, z_length = volume.shape_ in _dataset_generator_inference.py_. h=512 and w=512 always.
I have also checked the dimensions of the images from the dataset directly using _dicom = pydicom.dcmread(dicom_path)_ and then arr = _dicom.pixel_array_ . all the images have the dimension (512, 512).

3.I couldnt find official paper of the dataset describing the dataset. However, I found description about COCA in [this paper](https://link.springer.com/article/10.1007/s00330-025-11559-4). they have written: _In the Stanford-COCA dataset, most LDCT scans were acquired on GE (40%) and Siemens (57%) machines. For ECG-gated coronary calcium CT scans, Siemens machines were predominantly used (99%), with GE machines representing only 1%_.
so it seems that scanner manufacturer is not a problem because in the case of non-gated CTs scans have been done on GE and Siemens.

4.Also in the same above paper, I found this written in the acknowledgement of COCA dataset: _AIMI curated a publicly available imaging data repository containing clinical imaging and data from Stanford Health Care, the Stanford Children’s Hospital, the University Healthcare Alliance and Packard Children’s Health Alliance clinics provisioned for research use by the Stanford Medicine Research Data Repository (STARR)_. so it seems that it is not only the problem of gender skew, there is also age skew as veterans are older in age than children.

So overall, it seems that the lower performance (compared to veteran affaires datasets) is due to slice thickness and gender skew as you have said, and maybe also due to age skew.
Can you send me the link to the original paper that performed this work on COCA dataset?
Thank you very much for your help!

**Comment by Raffi-Hagopian** [comment]
Thank you for the follow-up! Here is the excellent Stanford paper (Eng. et al) on this topic, which I believe that public dataset derives from: [pubmed link](https://pubmed.ncbi.nlm.nih.gov/34075194/)

**Comment by koozalli** [comment]
thank you for sharing this paper!

I have found this written in the paper:
_. To increase the size of the internal training datasets, patients were eligible if a gated coronary CT was performed within no more than a 2-year time interval of the index chest CT exam, though the test set (n=42) comprised of only paired studies that were no more than one year apart (median time interval of 3.3 months). Rationale for restricting the test set to 1 year was based upon studies showing 8.3 Agatston unit change for patients with baseline scores of 1–100[63] and an overall low percentile (~5%) annualized change or median progression (29 Agatston units/year) in calcium score[63,64]._
and this written in the reference [[63](https://pmc.ncbi.nlm.nih.gov/articles/PMC4148074/)]in the paper:
_Among those with baseline CAC scores of 0, 0.001–99, 100–199, 200–299, and ≥300, median annual progression was 2.2, 8.3, 28.4, 44.0, and 103.3 units/year, respectively, for the imputed analysis and 0, 7.9, 27.7, 43.2, and 95.5 units/year, respectively, for the non-imputed analysis._

your model was tested on patients where the time difference between a non-gated and gated CT scans is less than one year. the test set that was sampled from the COCA dataset in stanford paper also comprised of only paired studies that were no more than one year apart. However, I have tested your model on all the stanford COCA non-gated dataset where the maximum allowed time difference between a gated and non-gated CT scans is 2 years.
so maybe one of the reasons for undercalling CACS might be that agatston scores might have increased in some patients with this >1 year period (more than the increase that might happen in less than one year), which contributed to the under-estimation of CAC on all COCA dataset compared to both your paper and stanford paper.

**Comment by Raffi-Hagopian** [comment]
I agree. Great discussion - hope the code was helpful for your work!
